Improve variable handling logic

dummyunit opened the pull request:
This series of commits bans most ways for a variable to change its type, which were discovered in #113, by adding checks that ensure that variable to be modified has same or compatible type.

It also fixes some others bugs/inconsistencies discovered along the way:
   - Subtraction operator wasn't handled properly in many cases. In operations where right operand was variable or explicit structure it worked like assignment. With this changes:
      1. Subtractions of structs are banned.
      2. Subtractions like `.String1 - .String2` work like `.String1 - "explicit string"`, removing occurrences of `.String2` from `.String1`.
      3. Subtractions like `.ArrayOfStrings - .String` work like `.ArrayOfStrings - "explicit string"`, removing elements that are equal to `.String` from `.ArrayOfStrings`.
      4. Added subtractions for integers (that was easier than adding error message :) ).
      5. Other subtractions are banned (bool and arrays).
   - Integer addition only worked when right operand was a variable. With this changes it will work with explicit values as well.

I have also changed bool value parsing to be more clear and not use cryptic constructions like `iter.ParseToNext( 'e' )`.

I must also mention that with these fixes code structured like that (which probably used by many people) now produces an error:
```
$ cat test.bff 
.EnabledX = { "Foo" "Bar" }
.Array = {}
ForEach( .X in .EnabledX )
{
        .Struct =
        [
                .A = .X
                .B = "-option $X$"
        ]
        ^Array + { .Struct }
}
$ fbuild -verbose -config test.bff
Registered <ArrayOfStrings> variable '.EnabledX' with 2 elements
Registered <ArrayOfStrings> variable '.Array' with 0 elements
Function call 'ForEach'
Freezing loop array '.EnabledX' of type <ArrayOfStrings>
Registered <string> variable '.A' with value 'Foo'
Registered <string> variable '.B' with value '-option Foo'
Registered <struct> variable '.Struct' with 2 members
/home/dummyunit/fastbuild/Code/test.bff(10,13): FASTBuild Error #1034 - Operation not supported: 'ArrayOfStrings' + 'Struct'.
        ^Array + { .Struct }
                   ^
                   \--here
```
So, if this will be merged, something must done to support empty arrays — either updated version of #110 or different internal type for fresh empty arrays (like I described in comment 3 in #110), or something else.

ffulin commented:
This change seems sensible to me.  Could you please resolve it against your previous commit?
Thanks again for you work on this!

dummyunit commented:
Done. Rebased on top of dev and adjusted some commits (mostly the first 2).

Also added one new commit (the last one) that fixes one rare case I didn't account for in #110: addition/subtraction/assignment of explicit string to empty ArrayOfStructs wasn't working, while the same with string _variable_ and empty ArrayOfStructs was.

ffulin commented:
I've integrated this into the Perforce mainline.  It will be in the next release (v0.91).  Thanks!
